source/mus_program_module.fpp

```diff
@@ -346,6 +366,156 @@ contains
 ! **************************************************************************** !
+  !> This routine does the main musubi computation loop for coupled LBM-DEM 
+  !! simulations of solid particles in a fluid
+  subroutine mus_solve_particles( scheme, geometry, params, control, &
+                                & solverData, adapt, particleGroup   )
+    ! ------------------------------------------------------------------------!
+    !> scheme type
+    type( mus_scheme_type ),    intent(inout) :: scheme
+    !> Treelmesh data
+    type( mus_geom_type ),      intent(inout) :: geometry
+    !> Global parameters
+    type( mus_param_type ),     intent(inout) :: params
+    !> control routine
+    type( mus_control_type ),   intent(in)    :: control
+    !> contains pointer to scheme, physics types
+    type( mus_varSys_solverData_type ), target :: solverData
+    !> mesh adaptation
+    type(tem_adapt_type),       intent(inout) :: adapt
+    !> object containing data for solid particles
+    type(mus_particle_group_type),    intent(inout) :: particleGroup
+    ! ------------------------------------------------------------------------!
+    integer :: minLevel, maxLevel
+    ! ------------------------------------------------------------------------!
+
+    call tem_startTimer( timerHandle =  mus_timerHandles%mainLoop )
+    minLevel = geometry%tree%global%minLevel
+    maxLevel = geometry%tree%global%maxLevel
+
+    ! --------------------------------------------------------------------------
+    ! Start main loop
+    mainloop: do
+
+      ! check if some action has to be taken based on timeControl:
+      ! tracking, global reduction operations, restart.
+      ! check flow status including perform checks
+      ! only at minLevel to complete one multilevel cycle.
+      call check_flow_status( scheme   = scheme,                    &
+        &                     general  = params%general,            &
+        &                     physics  = params%physics,            &
+        &                  mus_aborts  = params%mus_aborts,         &
+        &           restart_triggered  = params%restart_triggered,  &
+        &                    geometry  = geometry                   )
+
+      if( tem_status_run_end(params%general%simControl%status) .or.            &
+        & tem_status_run_terminate(params%general%simControl%status) ) then
+        exit mainLoop
+      end if
+
+      ! clear status bit
+      call tem_simControl_clearStat( me = params%general%simControl )
+
+      ! Iterate through the elements in a level-wise fashion
+      ! and advance the time steps
+      call control%do_computation_particles( scheme    = scheme,     &
+                                           & geometry  = geometry,   &
+                                           & params    = params,     &
+                                           & iLevel    = minLevel,   &
+                                           & particleGroup = particleGroup )
```

Comment:
If `do_computation_particles` is called *instead* of `do_computation` rather than in addition to it, there is no need for a separate interface and solve routine. Rather extend the `do_computation` interface with the `particleGroup` parameter and let the function pointer `do_computation` point to the respective function to use. Maybe it would also be possible to push `particleGroup` into `scheme`? Then you wouldn't need to change the interface here.